LIPs/LIP-snapshot_earnings.md

```diff
@@ -0,0 +1,146 @@
+
+## Specification
+
+The snapshot will be referenced under the form of a Merkle tree where each bottom-level will be a leaf consisting of `{delegator, rewards, fees}`. The root hash of this tree will be stored on-chain and a delegator can submit its owed earnings along with a proof to the `BondingManager` contract to claim its past earnings up until and including the snapshot round. 
+
+### Merkle Tree Generation
+
+1. Establish the snapshot round
+
+2. Get all delegator addresses
+
+3. For each delegator call 
+    - `pendingStake(delegator, snapshotRound)`
+    - `pendingFees(delegator, snapShotRound)`
+
+4. Sort the tree by `delegator` address
+
+A library such as [merkletreejs](https://github.com/miguelmota/merkletreejs) can be used to generate the tree and validate proofs off-chain before submitting a transaction
+
+### Snapshot Round 
+
+#35 provides a clear transitional round after which claiming earnings are a constant operation (O(1)). 
+
+It would thus make sense to create the snapshot for the round prior to the upgrade round for #35. This will turn claiming earnings into a constant operation accross the board. 
+
+### `MerkleSnapshot` Contract
+
+The `MerkleSnapshot` contract is a generic contract that allows to verify merkle proofs for any state snapshots. 
+
+#### State
+
+##### `snapshots` mapping
+
+A map of `ID => merkleRoot` allowing the contract to support multiple snapshots in the future if need be. 
+
+`ID` in most cases is the hash of `LIP-X`. 
+
+e.g. for this LIP `keccak256("LIP-52")`
+
+#### API
+
+##### `snapshot(bytes32 _ID) returns (bytes32 merkleRoot)`
+
+Auto-generated `external` getter that gets the specific merkle tree root hash for an snapshot `_ID`.
+
+##### `setSnapshot(bytes32 _ID, bytes32 _merkleRoot)` 
+
+Sets the root hash of a merkle tree representing a snapshot and updates the `snapshots` map for `_ID` with `_merkleRoot`. 
+
+This function reverts if the map entry already exists. 
+
+
+##### `verify(bytes32 _ID, bytes32[] _proof, bytes32 _leaf) external returns (bool)`
+
+Calls `MerkleProof.verify()` with the `_proof` and `_leaf` to verify it against the merkle root for a specific snapshot `_ID`.
+
+```
+bytes32 rootHash = snapshots[_ID];
+MerkleProof.verify(_proof, rootHash, _leaf)
+```
+
+#### `MerkleProof` Library
+
+`MerkleEarnings` will use `MerkleProof` contract from OpezenZepplin  to verify proofs: https://docs.openzeppelin.com/contracts/3.x/api/cryptography#MerkleProof
+
+This function exposes a single internal API 
+
+```
+verify(bytes32[] memory proof, bytes32 root, bytes32 leaf) internal pure returns (bool)
+```
+
+### `BondingManager` API changes
+
+#### `claimSnapShotEarnings(uint256 _rewards, uint256 _fees, bytes32[] memory _earningsProof)`
```

Comment:
I don't think the following is a strict requirement, but I was thinking about ways to improve the UX for a delegator and it would be nice if 1) claiming from the snapshot does not need to be a distinct action and a delegator can just do it when it also needs to perform a staking action 2) claiming from the snapshot and performing a staking action could  be done with a single tx. If both of these properties are achieved then a client that supports LIP-52 would not need to expose snapshot claiming to its user and it the snapshot claiming would happen in a tx that also performs a staking action requested by the user.

One way that these properties might be achieved could be to pass an optional tx data param for invoking another function after the snapshot claiming is complete. A rough example:

```solidity
function claimSnapshotEarnings(
    uint256 _rewards,
    uint256 _fees,
    bytes32[] memory _earningsProof,
    bytes _data
)
    external
    ...
{
    // Other code
    ...

    if (_data.length > 0) {
         // Delegatecall so msg.sender and the current call context is preserved
         // This could be a bond, unbond, rebond, etc.
        (bool success, bytes res) = address(this).delegatecall(_data);
        ...
    }
}
```

If achieving the aforementioned properties requires complex changes then I don't think we need to push for them, but if the change is simple (i.e. something along the lines of the example above) then it might be worth considering.

Reply:
Since this is not a formatting/correctness related comment this point can be discussed in the discussion issue.